SU2_CFD/include/numerics_simd/CNumericsSIMD.hpp

```diff
@@ -0,0 +1,711 @@
+  }
+  return vars;
+}
+
+template<size_t nVar, size_t nDim, class Field_t, class Gradient_t>
+FORCEINLINE VectorDbl<nVar> musclPointLimited(Int iPoint,
+                                              const VectorDbl<nDim>& vector_ij,
+                                              Double direction,
+                                              const Field_t& field,
+                                              const Field_t& limiter,
+                                              const Gradient_t& gradient) {
+  auto vars = gatherVariables<nVar>(iPoint, field);
+  auto itLim = limiter.template innerIter<nVar>(iPoint);
+  for (size_t iVar = 0; iVar < nVar; ++iVar) {
+    vars(iVar) += *(itLim++) * direction * dot(gradient.template innerIter<nVar,nDim>(iPoint,iVar), vector_ij);
+  }
+  return vars;
+}
+
+template<size_t nDim, size_t nVar, class Field_t, class Gradient_t>
+FORCEINLINE void musclEdgeLimited(Int iPoint,
+                                  Int jPoint,
+                                  const VectorDbl<nDim>& vector_ij,
+                                  const Field_t& field,
+                                  const Gradient_t& gradient,
+                                  VectorDbl<nVar>& vars_i,
+                                  VectorDbl<nVar>& vars_j) {
+  vars_i = gatherVariables<nVar>(iPoint, field);
+  vars_j = gatherVariables<nVar>(jPoint, field);
+
+  for (size_t iVar = 0; iVar < nVar; ++iVar) {
+    auto proj_i = dot(gradient.template innerIter<nVar,nDim>(iPoint,iVar), vector_ij);
+    auto proj_j = dot(gradient.template innerIter<nVar,nDim>(jPoint,iVar), vector_ij);
+    auto delta_ij = vars_j(iVar) - vars_i(iVar);
+    auto delta_ij_2 = pow(delta_ij,2);
+    /// TODO: Customize the limiter function
+    auto lim_i = (delta_ij_2 + 2*proj_i*delta_ij) / (4*pow(proj_i,2) + delta_ij_2 + EPS);
+    auto lim_j = (delta_ij_2 + 2*proj_j*delta_ij) / (4*pow(proj_j,2) + delta_ij_2 + EPS);
+    vars_i(iVar) += lim_i * proj_i;
+    vars_j(iVar) -= lim_j * proj_j;
+  }
+}
+
+template<class T>
+struct CPair {
+  T i, j;
+};
+
+template<size_t nDim>
+struct CCompressiblePrimitives {
+  enum : size_t {nVar = nDim+4};
+  VectorDbl<nVar> all;
+  FORCEINLINE Double& temperature() { return all(0); }
+  FORCEINLINE Double& pressure() { return all(nDim+1); }
+  FORCEINLINE Double& density() { return all(nDim+2); }
+  FORCEINLINE Double& enthalpy() { return all(nDim+3); }
+  FORCEINLINE Double& velocity(size_t iDim) { return all(iDim+1); }
+  FORCEINLINE const Double& temperature() const { return all(0); }
+  FORCEINLINE const Double& pressure() const { return all(nDim+1); }
+  FORCEINLINE const Double& density() const { return all(nDim+2); }
+  FORCEINLINE const Double& enthalpy() const { return all(nDim+3); }
+  FORCEINLINE const Double& velocity(size_t iDim) const { return all(iDim+1); }
+  FORCEINLINE const Double* velocity() const { return &velocity(0); }
+};
```

Comment:
I'm also trying to address the magic indexes (iDim+1) by defining specific types for primitive and conservative variables rather that using raw arrays.
Initially I thought of making `temperature, pressure, etc.` references pointing into the correct position of the storage vector (`all`)  but unfortunately that disables compiler generated constructors... so they are functions.